### PR TITLE
Increase polymorphism use where possible

### PR DESCRIPTION
**Pull request by BlairArchibald**
This cleans up the code by reducing the amount of "instanceof" etc.

It also clarifies what particular person types can do by, for example, only allowing Adults to have professions.

There is still some work to do in the handling of places to, for example, have randomSchool() etc functions.

### CODE REVIEW

File: src/main/java/uk/co/ramp/covid/simulation/place/CommunalPlace.java [hunk 1]
```diff
@@ -87,14 +91,14 @@ public boolean checkVisit(Person cPers, int time, int day, boolean clockdown) {
                             if (!nPers.getInfectionStatus()) {
                                 boolean infected = nPers.infChallenge(this.transProb * this.sDistance);
                                 if (infected) {
-                                    stats.infectedPlace(this, nPers);
+                                    registerInfection(stats, cPers);
```
Comment (johnnonweiler) [inline src/main/java/uk/co/ramp/covid/simulation/place/CommunalPlace.java:94]: Why has this has changed from nPers to cPers?

File: src/main/java/uk/co/ramp/covid/simulation/place/Household.java [hunk 2]
```diff
@@ -110,14 +110,15 @@ public boolean seedInfection() {
                             if (!nPers.getInfectionStatus()) {
                                 boolean infected = nPers.infChallenge(1);
                                 if (infected) {
-                                    stats.infectedHome(nPers);
+                                    stats.incInfectionsHome();
+                                    cPers.reportInfection(stats);
```
Comment (johnnonweiler) [inline src/main/java/uk/co/ramp/covid/simulation/place/Household.java:114]: Similarly here - why the change from nPers to cPers?

File: src/main/java/uk/co/ramp/covid/simulation/place/CommunalPlace.java (hunk 1 above)
Comment (BlairArchibald) [reply src/main/java/uk/co/ramp/covid/simulation/place/CommunalPlace.java:94]: Good spots, looks like bugs - I'll fix this up

File: src/main/java/uk/co/ramp/covid/simulation/population/Pensioner.java [hunk 3]
```diff
@@ -1,5 +1,25 @@
...
+        s.incInfectionsPensioner();
+    }
+
+    @Override
+    public void reportDeath(DailyStats s) {
+        s.incDeathsPensioner();
+    }
+
+    @Override
+    public void allocateCommunalPlace(Population p) {}
```
Comment (bobtaylor12) [inline src/main/java/uk/co/ramp/covid/simulation/population/Pensioner.java:19]: Do pensioners need to be allocated to communal places?

File: src/main/java/uk/co/ramp/covid/simulation/population/Pensioner.java [hunk 4]
```diff
@@ -1,5 +1,25 @@
 package uk.co.ramp.covid.simulation.population;
 
+import uk.co.ramp.covid.simulation.CovidParameters;
+import uk.co.ramp.covid.simulation.DailyStats;
+
 public class Pensioner extends Person {
```
Comment (bobtaylor12) [inline src/main/java/uk/co/ramp/covid/simulation/population/Pensioner.java:6]: Would it be more logical for pensioners to extend Adult?

File: src/main/java/uk/co/ramp/covid/simulation/population/Person.java [hunk 5]
```diff
@@ -108,20 +51,24 @@ public void setRecovered(boolean recovered) {
         this.recovered = recovered;
     }
 
-    public int getMIndex() {
-        return this.mIndex;
+    public CommunalPlace getPrimaryCommunalPlace() {
+        return primaryPlace;
```
Comment (bobtaylor12) [inline src/main/java/uk/co/ramp/covid/simulation/population/Person.java:55]: I would prefer to use 'this.primaryPlace'

File: src/main/java/uk/co/ramp/covid/simulation/Covid.java [hunk 6]
```diff
@@ -69,24 +69,16 @@ public boolean isDead() {
...
-        if ((this.ccase instanceof Adult)
-                && rng.nextUniform(0, 1) > CovidParameters.get().getAdultProgressionPhase2()) {
-            this.p2 = 0;
-        }
-        if ((this.ccase instanceof Pensioner)
-                && rng.nextUniform(0, 1) > CovidParameters.get().getPensionerProgressionPhase2()) {
-            this.p2 = 0;
-        }
+        latentPeriod = (int) rng.nextPoisson(meanLatentPeriod);
+        latentPeriod = (int) rng.nextPoisson(meanAsymptomaticPeriod);
```
[the remark this answers is not in the document]
Comment (BlairArchibald) [reply src/main/java/uk/co/ramp/covid/simulation/Covid.java:73]: Good spot - looks like this is set in the constructor anyway; I'll just remove this.

File: src/main/java/uk/co/ramp/covid/simulation/population/Pensioner.java (hunk 3 above)
Comment (BlairArchibald) [reply src/main/java/uk/co/ramp/covid/simulation/population/Pensioner.java:19]: Not currently. This is used to assign people workplaces/schools/nurseries - Pensioners currently only go to restaurants/shopping (which are handled separately). Probably useful to have this hook for future extensions to to keep the API uniform.

File: src/main/java/uk/co/ramp/covid/simulation/population/Pensioner.java (hunk 4 above)
Comment (BlairArchibald) [reply src/main/java/uk/co/ramp/covid/simulation/population/Pensioner.java:6]: Probably not since we don't want to inherit professions etc. They are treated pretty differently throughout.

File: src/main/java/uk/co/ramp/covid/simulation/population/Person.java (hunk 5 above)
Comment (BlairArchibald) [reply src/main/java/uk/co/ramp/covid/simulation/population/Person.java:55]: Sure I can make this change.

File: src/main/java/uk/co/ramp/covid/simulation/Covid.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (BlairArchibald) [reply src/main/java/uk/co/ramp/covid/simulation/Covid.java:73]: Good spot - I'll fix this!
